File: sdk/hypercli/claw.py
"""
HyperClaw API client

Provides access to the HyperClaw inference API for AI agents.
Uses the official OpenAI Python client for chat completions.
"""
from dataclasses import dataclass
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from .http import HTTPClient

# OpenAI client is optional - only needed for chat
try:
    from openai import OpenAI
    OPENAI_AVAILABLE = True
except ImportError:
    OPENAI_AVAILABLE = False


@dataclass
class ClawKey:
    """HyperClaw API key with subscription details."""
    key: str
    plan_id: str
    expires_at: datetime
    tpm_limit: int
    rpm_limit: int
    user_id: Optional[str] = None
    
    @classmethod
    def from_dict(cls, data: dict) -> "ClawKey":
        expires = data.get("expires_at", "")
        if isinstance(expires, str):
            expires = datetime.fromisoformat(expires.replace("Z", "+00:00"))
        return cls(
            key=data["key"],
            plan_id=data["plan_id"],
            expires_at=expires,
            tpm_limit=data.get("tpm_limit", 0),
            rpm_limit=data.get("rpm_limit", 0),
            user_id=data.get("user_id"),
        )


@dataclass
class ClawPlan:
    """HyperClaw subscription plan."""
    id: str
    name: str
    price_usd: float
    tpm_limit: int
    rpm_limit: int
    
    @classmethod
    def from_dict(cls, data: dict) -> "ClawPlan":
        return cls(
            id=data["id"],
            name=data["name"],
            price_usd=data["price_usd"],
            tpm_limit=data["tpm_limit"],
            rpm_limit=data["rpm_limit"],
        )


@dataclass  
class ClawModel:
    """Available model on HyperClaw."""
    id: str
    name: str
    context_length: int
    supports_vision: bool = False
    supports_function_calling: bool = False
    supports_tool_choice: bool = False
    
    @classmethod
    def from_dict(cls, data: dict) -> "ClawModel":
        caps = data.get("capabilities", {})
        return cls(
            id=data["id"],
            name=data.get("name", data["id"]),
            context_length=data.get("context_length", 0),
            supports_vision=caps.get("supports_vision", False),
            supports_function_calling=caps.get("supports_function_calling", False),
            supports_tool_choice=caps.get("supports_tool_choice", False),
        )


class Claw:
    """
    HyperClaw API Client
    
    Provides access to HyperClaw inference endpoints using the OpenAI Python client.
    
    Usage:
        from hypercli import HyperCLI
        
        client = HyperCLI(claw_api_key="sk-...")
        
        # Get OpenAI client for chat
        openai = client.claw.openai
        response = openai.chat.completions.create(
            model="kimi-k2.5",
            messages=[{"role": "user", "content": "Hello!"}]
        )
        
        # Or use the convenience wrapper
        response = client.claw.chat(
            model="kimi-k2.5",
            messages=[{"role": "user", "content": "Hello!"}]
        )
        
        # Vision
        response = client.claw.chat(
            model="kimi-k2.5",
            messages=[{
                "role": "user",
                "content": [
                    {"type": "text", "text": "What's in this image?"},
                    {"type": "image_url", "image_url": {"url": "data:image/jpeg;base64,..."}}
                ]
            }]
        )
        
        # Function calling
        response = client.claw.chat(
            model="kimi-k2.5",
            messages=[...],
            tools=[{"type": "function", "function": {...}}]
        )
    """
    
    # Default HyperClaw API base URL
    CLAW_API_BASE = "https://api.hyperclaw.app/v1"
    DEV_API_BASE = "https://dev-api.hyperclaw.app/v1"
    
    def __init__(self, http: HTTPClient, claw_api_key: str = None, dev: bool = False):
        """
        Initialize HyperClaw client.
        
        Args:
            http: HTTPClient for making requests (used for non-OpenAI endpoints)
            claw_api_key: Optional separate API key for HyperClaw
            dev: Use dev API endpoint
        """
        self._http = http
        self._api_key = claw_api_key or http.api_key
        self._dev = dev
        self._base_url = self.DEV_API_BASE if dev else self.CLAW_API_BASE
        self._openai = None
    
    @property
    def openai(self) -> "OpenAI":
        """
        Get OpenAI client configured for HyperClaw.
        
        Returns:
            OpenAI client instance
            
        Raises:
            ImportError: If openai package is not installed
        """
        if not OPENAI_AVAILABLE:
            raise ImportError(
                "OpenAI package required for chat. Install with: pip install openai"
            )
        
        if self._openai is None:
            self._openai = OpenAI(
                api_key=self._api_key,
                base_url=self._base_url,
            )
        return self._openai
    
    def chat(
        self,
        model: str,
        messages: List[Dict],
        temperature: float = None,
        max_tokens: int = None,
        tools: List[Dict] = None,
        tool_choice: Union[str, Dict] = None,
        stream: bool = False,
        **kwargs
    ):
        """
        Create a chat completion using the OpenAI client.
        
        Args:
            model: Model ID (e.g., "kimi-k2.5")
            messages: List of message dicts
            temperature: Sampling temperature (0-2)
            max_tokens: Maximum tokens to generate
            tools: List of tool definitions for function calling
            tool_choice: Tool choice mode ("auto", "none", or specific tool)
            stream: Whether to stream the response
            **kwargs: Additional parameters passed to the API
            
        Returns:
            OpenAI ChatCompletion object
        """
        params = {
            "model": model,
            "messages": messages,
            **kwargs
        }
        
        if temperature is not None:
            params["temperature"] = temperature
        if max_tokens is not None:
            params["max_tokens"] = max_tokens
        if tools:
            params["tools"] = tools
        if tool_choice:
            params["tool_choice"] = tool_choice
        if stream:
            params["stream"] = stream
        
        return self.openai.chat.completions.create(**params)
    
    def models(self) -> List[ClawModel]:
        """
        List available models.
        
        Returns:
            List of ClawModel objects
        """
        response = self.openai.models.list()
        return [
            ClawModel.from_dict({
                "id": m.id,
                "name": getattr(m, "name", m.id),
                "context_length": getattr(m, "context_length", 0),
                "capabilities": getattr(m, "capabilities", {}),
            })
            for m in response.data
        ]
    
    def _api_base_without_v1(self) -> str:
        """Get API base URL without /v1 suffix."""
        return self._base_url.replace("/v1", "")
    
    def key_status(self) -> ClawKey:
        """
        Get current API key status and subscription details.
        
        Returns:
            ClawKey object with subscription info
        """
        r = self._http._session.get(
            f"{self._api_base_without_v1()}/api/keys/status",
            headers={"Authorization": f"Bearer {self._api_key}"},
        )
        r.raise_for_status()
        return ClawKey.from_dict(r.json())
    
    def plans(self) -> List[ClawPlan]:
        """
        List available subscription plans.
        
        Returns:
            List of ClawPlan objects
        """
        r = self._http._session.get(
            f"{self._api_base_without_v1()}/api/plans",
            headers={"Authorization": f"Bearer {self._api_key}"},
        )
        r.raise_for_status()
        data = r.json()
        return [ClawPlan.from_dict(p) for p in data.get("plans", [])]
    
    def discovery_health(self) -> Dict[str, Any]:
        """
        Get discovery service health status.
        
        Returns:
            Dict with hosts_total, hosts_healthy, fallbacks_active
        """
        r = self._http._session.get(
            f"{self._api_base_without_v1()}/discovery/health",
        )
        r.raise_for_status()
        return r.json()
    
    def discovery_config(self, api_key: str = None) -> Dict[str, Any]:
        """
        Get discovery service configuration (requires API key).
        
        Args:
            api_key: Backend API key (not the user's claw key)
            
        Returns:
            Dict with hosts, fallbacks, config
        """
        headers = {}
        if api_key:
            headers["X-API-KEY"] = api_key
        
        r = self._http._session.get(
            f"{self._api_base_without_v1()}/discovery/config",
            headers=headers,
        )
        r.raise_for_status()
        return r.json()

File: ts-sdk/src/claw.ts
/**
 * HyperClaw API client - AI agent inference using OpenAI-compatible API
 * 
 * Note: OpenAI client integration is not included in this SDK.
 * Use the OpenAI Node.js SDK directly with HyperClaw endpoints.
 */
import type { HTTPClient } from './http.js';

export interface ClawKey {
  key: string;
  planId: string;
  expiresAt: Date;
  tpmLimit: number;
  rpmLimit: number;
  userId: string | null;
}

export interface ClawPlan {
  id: string;
  name: string;
  priceUsd: number;
  tpmLimit: number;
  rpmLimit: number;
}

export interface ClawModel {
  id: string;
  name: string;
  contextLength: number;
  supportsVision: boolean;
  supportsFunctionCalling: boolean;
  supportsToolChoice: boolean;
}

function clawKeyFromDict(data: any): ClawKey {
  let expiresAt: Date;
  if (typeof data.expires_at === 'string') {
    expiresAt = new Date(data.expires_at.replace('Z', '+00:00'));
  } else {
    expiresAt = new Date(data.expires_at);
  }

  return {
    key: data.key,
    planId: data.plan_id,
    expiresAt,
    tpmLimit: data.tpm_limit || 0,
    rpmLimit: data.rpm_limit || 0,
    userId: data.user_id || null,
  };
}

function clawPlanFromDict(data: any): ClawPlan {
  return {
    id: data.id,
    name: data.name,
    priceUsd: data.price_usd,
    tpmLimit: data.tpm_limit,
    rpmLimit: data.rpm_limit,
  };
}

function clawModelFromDict(data: any): ClawModel {
  const caps = data.capabilities || {};
  return {
    id: data.id,
    name: data.name || data.id,
    contextLength: data.context_length || 0,
    supportsVision: caps.supports_vision || false,
    supportsFunctionCalling: caps.supports_function_calling || false,
    supportsToolChoice: caps.supports_tool_choice || false,
  };
}

/**
 * HyperClaw API Client
 * 
 * For chat completions, use the OpenAI Node.js SDK directly:
 * 
 * ```typescript
 * import OpenAI from 'openai';
 * 
 * const openai = new OpenAI({
 *   apiKey: client.claw.apiKey,
 *   baseURL: client.claw.baseUrl,
 * });
 * 
 * const response = await openai.chat.completions.create({
 *   model: 'kimi-k2.5',
 *   messages: [{ role: 'user', content: 'Hello!' }],
 * });
 * ```
 */
export class Claw {
  static readonly CLAW_API_BASE = 'https://api.hyperclaw.app/v1';
  static readonly DEV_API_BASE = 'https://dev-api.hyperclaw.app/v1';

  public readonly apiKey: string;
  public readonly baseUrl: string;

  constructor(
    private http: HTTPClient,
    clawApiKey?: string,
    dev: boolean = false
  ) {
    this.apiKey = clawApiKey || http['apiKey'];
    this.baseUrl = dev ? Claw.DEV_API_BASE : Claw.CLAW_API_BASE;
  }

  private get baseUrlWithoutV1(): string {
    return this.baseUrl.replace('/v1', '');
  }

  /**
   * Get current API key status and subscription details
   */
  async keyStatus(): Promise<ClawKey> {
    const response = await fetch(`${this.baseUrlWithoutV1}/api/keys/status`, {
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to get key status: ${response.statusText}`);
    }

    const data = await response.json();
    return clawKeyFromDict(data);
  }

  /**
   * List available subscription plans
   */
  async plans(): Promise<ClawPlan[]> {
    const response = await fetch(`${this.baseUrlWithoutV1}/api/plans`, {
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to get plans: ${response.statusText}`);
    }

    const data: any = await response.json();
    return (data.plans || []).map(clawPlanFromDict);
  }

  /**
   * List available models
   */
  async models(): Promise<ClawModel[]> {
    const response = await fetch(`${this.baseUrl}/models`, {
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to get models: ${response.statusText}`);
    }

    const data: any = await response.json();
    return (data.data || []).map((m: any) => clawModelFromDict({
      id: m.id,
      name: m.name || m.id,
      context_length: m.context_length || 0,
      capabilities: m.capabilities || {},
    }));
  }

  /**
   * Get discovery service health status
   */
  async discoveryHealth(): Promise<{
    hostsTotal: number;
    hostsHealthy: number;
    fallbacksActive: number;
  }> {
    const response = await fetch(`${this.baseUrlWithoutV1}/discovery/health`);

    if (!response.ok) {
      throw new Error(`Failed to get discovery health: ${response.statusText}`);
    }

    return (await response.json()) as any;
  }

  /**
   * Get discovery service configuration (requires API key)
   */
  async discoveryConfig(apiKey?: string): Promise<any> {
    const headers: Record<string, string> = {};
    if (apiKey) {
      headers['X-API-KEY'] = apiKey;
    }

    const response = await fetch(`${this.baseUrlWithoutV1}/discovery/config`, {
      headers,
    });

    if (!response.ok) {
      throw new Error(`Failed to get discovery config: ${response.statusText}`);
    }

    return await response.json();
  }
}
